File: src/live2d_llm_expression/engine.py
from __future__ import annotations

from pathlib import Path

from live2d_llm_expression.emotion.schema import (
    EmotionIntent,
    EmotionName,
    SpecialExpressionName,
)
from live2d_llm_expression.llm.analyzer import (
    EmotionAnalyzer,
    MockEmotionAnalyzer,
    OpenAICompatibleAnalyzer,
)
from live2d_llm_expression.mapper.rule_mapper import RuleBasedExpressionMapper
from live2d_llm_expression.mapper.safety_clamp import clamp_params
from live2d_llm_expression.mapper.timeline_builder import build_timeline
from live2d_llm_expression.profile.model_profile import CharacterProfile
from live2d_llm_expression.profile.profile_builder import build_character_profile
from live2d_llm_expression.runtime.output import ExpressionResult, TimelineExpressionResult
from live2d_llm_expression.scanner import scan_live2d_resources


class Live2DExpressionEngine:
    def __init__(
        self,
        profile: CharacterProfile,
        mapper: RuleBasedExpressionMapper | None = None,
        analyzer: EmotionAnalyzer | None = None,
    ) -> None:
        self.profile = profile
        self.mapper = mapper or RuleBasedExpressionMapper()
        self.analyzer = analyzer or MockEmotionAnalyzer()

    @classmethod
    def from_directory(
        cls,
        root_dir: str | Path,
        *,
        analyzer: EmotionAnalyzer | None = None,
    ) -> "Live2DExpressionEngine":
        resources = scan_live2d_resources(root_dir)
        return cls(build_character_profile(resources), analyzer=analyzer)

    @classmethod
    def from_directory_with_env_analyzer(
        cls,
        root_dir: str | Path,
    ) -> "Live2DExpressionEngine":
        return cls.from_directory(root_dir, analyzer=OpenAICompatibleAnalyzer.from_env())

    def generate_by_emotion(
        self,
        emotion: EmotionName,
        *,
        intensity: float = 0.5,
        gaze: str | None = None,
        head: str | None = None,
        eyes: str | None = None,
        brows: str | None = None,
        mouth: str | None = None,
        special_expression: SpecialExpressionName | None = None,
        duration_ms: int = 1200,
    ) -> ExpressionResult:
        intent = EmotionIntent(
            emotion=emotion,
            intensity=intensity,
            gaze=gaze,
            head=head,
            eyes=eyes,
            brows=brows,
            mouth=mouth,
            special_expression=special_expression,
            duration_ms=duration_ms,
        )
        return self.generate_from_intent(intent)

    def generate_from_intent(self, intent: EmotionIntent) -> ExpressionResult:
        raw_params = self.mapper.map_intent(intent)
        params, warnings = clamp_params(raw_params, self.profile)
        return ExpressionResult(
            emotion=intent.emotion,
            intensity=intent.intensity,
            duration_ms=intent.duration_ms,
            params=params,
            source_intent=intent,
            warnings=warnings,
        )

    def generate_from_text(self, text: str) -> ExpressionResult:
        return self.generate_from_intent(self.analyzer.analyze(text))

    def generate_timeline_by_emotion(
        self,
        emotion: EmotionName,
        *,
        intensity: float = 0.5,
        gaze: str | None = None,
        head: str | None = None,
        eyes: str | None = None,
        brows: str | None = None,
        mouth: str | None = None,
        special_expression: SpecialExpressionName | None = None,
        duration_ms: int = 1200,
    ) -> TimelineExpressionResult:
        intent = EmotionIntent(
            emotion=emotion,
            intensity=intensity,
            gaze=gaze,
            head=head,
            eyes=eyes,
            brows=brows,
            mouth=mouth,
            special_expression=special_expression,
            duration_ms=duration_ms,
        )
        return self.generate_timeline_from_intent(intent)

    def generate_timeline_from_intent(
        self,
        intent: EmotionIntent,
    ) -> TimelineExpressionResult:
        neutral_intent = EmotionIntent(
            emotion="neutral",
            intensity=1.0,
            duration_ms=intent.duration_ms,
        )
        neutral_result = self.generate_from_intent(neutral_intent)
        target_result = self.generate_from_intent(intent)
        return build_timeline(
            intent,
            neutral_params=neutral_result.params,
            target_params=target_result.params,
            warnings=neutral_result.warnings + target_result.warnings,
        )

    def generate_timeline_from_text(self, text: str) -> TimelineExpressionResult:
        return self.generate_timeline_from_intent(self.analyzer.analyze(text))

File: src-ts/engine.ts
import { MockEmotionAnalyzer } from "./analyzer.js";
import { materializeEmotionSignalPreset } from "./emotion-signal.js";
import { normalizeIntent } from "./intent.js";
import { clampParams, mapIntentToParams } from "./mapper.js";
import {
  applyNaturalParameterMotion,
  createNaturalMotionPlan,
  stabilizeNaturalMotionKeyframes,
  type NaturalMotionOptions,
} from "./natural-motion.js";
import { buildMotionCapability } from "./motion-capability.js";
import { buildParameterManifest } from "./parameter-manifest.js";
import { buildCharacterProfile } from "./profile.js";
import { createResourceSetFromModel3Url, createResourceSetFromUrls } from "./resources.js";
import { buildTimeline } from "./timeline.js";
import type {
  CharacterProfile,
  EmotionAnalyzer,
  EmotionIntent,
  EmotionName,
  ExpressionResult,
  Live2DMotionCapability,
  Live2DModel3Urls,
  Live2DParameterManifest,
  Live2DResourceSet,
  Live2DResourceUrls,
  SpecialExpressionName,
  TimelineExpressionResult,
} from "./types.js";

export interface Live2DExpressionEngineOptions {
  analyzer?: EmotionAnalyzer;
}

export interface GenerateEmotionOptions {
  intensity?: number;
  gaze?: string | null;
  head?: string | null;
  eyes?: string | null;
  brows?: string | null;
  mouth?: string | null;
  specialExpression?: SpecialExpressionName | null;
  durationMs?: number;
}

export interface GenerateNaturalTimelineOptions extends GenerateEmotionOptions, NaturalMotionOptions {}

export class Live2DExpressionEngine {
  readonly profile: CharacterProfile;
  readonly analyzer: EmotionAnalyzer;

  constructor(profile: CharacterProfile, options: Live2DExpressionEngineOptions = {}) {
    this.profile = profile;
    this.analyzer = options.analyzer ?? new MockEmotionAnalyzer();
  }

  getParameterManifest(): Live2DParameterManifest {
    return buildParameterManifest(this.profile);
  }

  getMotionCapability(): Live2DMotionCapability {
    return buildMotionCapability(this.profile);
  }

  static async fromResourceSet(
    resources: Live2DResourceSet,
    options: Live2DExpressionEngineOptions & { fetcher?: typeof fetch } = {},
  ): Promise<Live2DExpressionEngine> {
    return new Live2DExpressionEngine(await buildCharacterProfile(resources, options), options);
  }

  static async fromUrls(
    resources: Live2DResourceUrls,
    options: Live2DExpressionEngineOptions & { fetcher?: typeof fetch } = {},
  ): Promise<Live2DExpressionEngine> {
    return Live2DExpressionEngine.fromResourceSet(createResourceSetFromUrls(resources), options);
  }

  static async fromModel3Url(
    resources: Live2DModel3Urls,
    options: Live2DExpressionEngineOptions & { fetcher?: typeof fetch } = {},
  ): Promise<Live2DExpressionEngine> {
    return Live2DExpressionEngine.fromResourceSet(
      await createResourceSetFromModel3Url(resources, options.fetcher),
      options,
    );
  }

  static async fromNodeDirectory(
    rootDir: string,
    options: Live2DExpressionEngineOptions = {},
  ): Promise<Live2DExpressionEngine> {
    const { scanLive2DResources, readNodeJsonResource } = await import("./node-resources.js");
    const resources = await scanLive2DResources(rootDir);
    return new Live2DExpressionEngine(
      await buildCharacterProfile(resources, {
        jsonLoader: (href) => readNodeJsonResource(href),
      }),
      options,
    );
  }

  generateByEmotion(
    emotion: EmotionName,
    options: GenerateEmotionOptions = {},
  ): ExpressionResult {
    return this.generateFromIntent({ emotion, ...options });
  }

  generateFromIntent(intentInput: EmotionIntent): ExpressionResult {
    const intent = normalizeIntent(
      intentInput.presetId ? materializeEmotionSignalPreset(intentInput) : intentInput,
    );
    const { params, warnings } = clampParams(mapIntentToParams(intent, this.profile), this.profile);
    return {
      emotion: intent.emotion,
      tone: intent.tone,
      intensity: intent.intensity,
      durationMs: intent.durationMs,
      params,
      sourceIntent: intent,
      warnings,
    };
  }

  async generateFromText(text: string): Promise<ExpressionResult> {
    return this.generateFromIntent(await this.analyzer.analyze(text));
  }

  generateTimelineByEmotion(
    emotion: EmotionName,
    options: GenerateEmotionOptions = {},
  ): TimelineExpressionResult {
    return this.generateTimelineFromIntent({ emotion, ...options });
  }

  generateTimelineFromIntent(intentInput: EmotionIntent): TimelineExpressionResult {
    const intent = normalizeIntent(intentInput);
    const neutral = this.generateFromIntent({ emotion: "neutral", intensity: 1, durationMs: intent.durationMs });
    const target = this.generateFromIntent(intent);
    return buildTimeline({
      intent,
      neutralParams: neutral.params,
      targetParams: target.params,
      warnings: [...neutral.warnings, ...target.warnings],
    });
  }

  async generateTimelineFromText(text: string): Promise<TimelineExpressionResult> {
    return this.generateTimelineFromIntent(await this.analyzer.analyze(text));
  }

  generateNaturalTimelineByEmotion(
    emotion: EmotionName,
    options: GenerateNaturalTimelineOptions = {},
  ): TimelineExpressionResult {
    return this.generateNaturalTimelineFromIntent({ emotion, ...options }, options);
  }

  generateNaturalTimelineFromIntent(
    intentInput: EmotionIntent,
    options: NaturalMotionOptions = {},
  ): TimelineExpressionResult {
    const motionOptions = {
      ...options,
      motionCapability: options.motionCapability ?? this.getMotionCapability(),
    };
    const plan = createNaturalMotionPlan(intentInput, motionOptions);
    const warnings = new Set<string>();
    const keyframes = plan.steps.map((step) => {
      const result = this.generateFromIntent(step.intent);
      result.warnings.forEach((warning) => warnings.add(warning));
      const paramsWithMotion = applyNaturalParameterMotion(result.params, step, plan, motionOptions);
      const clamped = clampParams(paramsWithMotion, this.profile);
      clamped.warnings.forEach((warning) => warnings.add(warning));
      return {
        t: step.t,
        phase: step.phase,
        params: clamped.params,
      };
    });

    return {
      emotion: plan.intent.emotion,
      tone: plan.intent.tone,
      intensity: plan.intent.intensity,
      durationMs: plan.durationMs,
      keyframes: stabilizeNaturalMotionKeyframes(keyframes, motionOptions),
      warnings: [...warnings],
    };
  }

  async generateNaturalTimelineFromText(
    text: string,
    options: NaturalMotionOptions = {},
  ): Promise<TimelineExpressionResult> {
    return this.generateNaturalTimelineFromIntent(await this.analyzer.analyze(text), options);
  }
}
